### PR TITLE
ENT-3756: cf-agent intermittently hangs on windows

### PR DESCRIPTION
**Pull request by craigcomstock**
merge together:
cfengine/core/pull/3180
cfengine/enterprise/pull/429

### CODE REVIEW

**Comment by craigcomstock** [comment]
made this forward port of my 3.7.x changes so I can hopefully see travis codecov report. Needed to be done anyway.

File: HACKING.md [hunk 1]
```diff
@@ -386,6 +386,28 @@ Tip: In order to trigger assert() calls in the code, build with
...
+uses to generate lcov.info and the HTML report in the coverage-html directory.
+
+Many IDEs and editors expect a <root>/coverage/lcov.info summary of coverage
+information. After running `make check` you can run `make coverage` and 
+generate this lcov.info summary for use with other tools. If you wish to only
+run a few tests which will add to coverage data you can update lcov.info with
+`make collect-coverage` which will only collect coverage data, not compile or
+run any tests.
+
+For the atom editor, install the package atom-lcov-info.
```
Comment (olehermanse) [inline HACKING.md]: This should be in a separate PR.

File: tests/unit/Makefile.am [hunk 2]
```diff
@@ -111,7 +111,6 @@ check_PROGRAMS = \
 	regex_test \
 	alloc_test \
 	string_writer_test \
-	file_writer_test \
 	xml_writer_test \
```
Comment (craigcomstock) [inline tests/unit/Makefile.am]: TODO, need to re-investigate whether we can remove/include file_writer_test

File: tests/unit/Makefile.am [hunk 3]
```diff
@@ -180,6 +179,8 @@ check_PROGRAMS = \
 	new_packages_promise_test \
 	iteration_test
 
+# todo fix this one: duplicate def from gcov_stub.c|h file_writer_test
+
 if HAVE_AVAHI_CLIENT
```
Comment (craigcomstock) [inline tests/unit/Makefile.am]: investigate this todo comment

File: tests/unit/Makefile.am [hunk 4]
```diff
@@ -254,12 +255,14 @@ buffer_test_SOURCES = buffer_test.c ../../libutils/buffer.c
...
 csv_parser_test_SOURCES = csv_parser_test.c ../../libutils/csv_parser.c
-csv_parser_test_LDADD = libtest.la ../../libutils/libutils.la
 
 regex_test_SOURCES = regex_test.c ../../libpromises/match_scope.c
 
 ipaddress_test_SOURCES = ipaddress_test.c
 
+#TODO remove this if tests pass
+#item_test_SOURCES = item_test.c
+
```
Comment (craigcomstock) [inline tests/unit/Makefile.am]: check on item_test

File: tests/unit/Makefile.am [hunk 5]
```diff
@@ -331,33 +330,28 @@ file_lib_test_SOURCES = file_lib_test.c \
...
+connection_management_test_LDADD = ../../cf-serverd/libcf-serverd.la ../../libutils/libutils.la ../../libpromises/libpromises.la libtest.la
 
-rlist_test_SOURCES = rlist_test.c \
-	../../libpromises/rlist.c
-rlist_test_LDADD = libtest.la ../../libpromises/libpromises.la
+rlist_test_SOURCES = rlist_test.c
+# TODO remove this if tests pass
+#\
+#	../../libpromises/rlist.c
 # Workaround for object file basename conflicts, search the web for
```
Comment (craigcomstock) [inline tests/unit/Makefile.am]: check on rlist_test success/fail

File: libutils/logging.c [hunk 6]
```diff
@@ -216,6 +218,11 @@ static void LogToConsole(const char *msg, LogLevel level, bool color)
 
     fprintf(stdout, "%8s: %s\n", LogLevelToString(level), msg);
 
+    if (level == LOG_LEVEL_DEBUG)
+    {
+        fflush(stdout);
+    }
```
Comment (craigcomstock) [inline libutils/logging.c]: @vpodzime do you think I should include an fsync() call here just to make more certain the logs are written every time Log() is called? (from an older PR which I have now closed)

File: cf-monitord/env_monitor.c [hunk 7]
```diff
@@ -317,11 +317,15 @@ void MonitorStartServer(EvalContext *ctx, const Policy *policy)
 
     MonNetworkSnifferOpen();
 
+printf("Before IsPendingTermination() while loop...\n");
+
     while (!IsPendingTermination())
```
Comment (craigcomstock) [inline cf-monitord/env_monitor.c]: cleanup

File: cf-monitord/env_monitor.c [hunk 8]
```diff
@@ -317,11 +317,15 @@ void MonitorStartServer(EvalContext *ctx, const Policy *policy)
...
     {
         GetQ(ctx, policy);
         snprintf(timekey, sizeof(timekey), "%s", GenTimeKey(time(NULL)));
         averages = EvalAvQ(ctx, timekey);
+// TODO if (averages == NULL) { break; }
+// why exit in EvalAvQ() and not do PolicyDestroy() and YieldCurrentLock()?
         LeapDetection();
```
Comment (craigcomstock) [inline cf-monitord/env_monitor.c]: cleanup

File: cf-monitord/env_monitor.c [hunk 9]
```diff
@@ -332,6 +336,8 @@ void MonitorStartServer(EvalContext *ctx, const Policy *policy)
         ITER++;
     }
 
+printf("Outside of IsPendingTermination() while loop\n");
+
     PolicyDestroy(monitor_cfengine_policy);
```
Comment (craigcomstock) [inline cf-monitord/env_monitor.c]: cleanup

File: cf-monitord/env_monitor.c [hunk 10]
```diff
@@ -374,10 +380,20 @@ static Averages EvalAvQ(EvalContext *ctx, char *t)
...
+    if (!NDEBUG && getenv("DEBUG_TEST_EVALAVQ_EXITS"))
+    {
+        Log(LOG_LEVEL_DEBUG, "env has DEBUG_TEST_EVALAVQ_EXITS so exiting as if GetCurrentAverages() returned NULL.");
+        DoCleanupAndExit(EXIT_FAILURE);
+    }
+
+
+// TODO why not just return NULL and have calling code handle the exit?
+
     if ((lastweek_vals = GetCurrentAverages(t)) == NULL)
```
Comment (craigcomstock) [inline cf-monitord/env_monitor.c]: remove, I found a way to do this that is not as intrusive

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/1019/)

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/1021/)

File: libutils/logging.c (hunk 6 above)
Comment (vpodzime) [reply libutils/logging.c]: Yeah, I don't think it is a big deal if it only happens with `LOG_LEVEL_DEBUG`.

File: tests/acceptance/15_control/02_agent/allclassesreport.cf [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+bundle agent check
+{
+  methods:
+      "Pass/Fail" usebundle => dcs_passif_fileexists( "$(sys.statedir)/allclasses.txt",
+                                                      $(this.promise_filename));
+
+  reports:
+      DEBUG|EXTRA::
```
Comment (vpodzime) [inline tests/acceptance/15_control/02_agent/allclassesreport.cf]: Wrong indentation.

File: tests/acceptance/15_control/02_agent/no-policy.cf [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+# for this one we expect failure
+    "$(sys.cf_agent) -f ./does-not-exist.cf";
+}
+
+body classes kept
+{
+  promise_kept => { "ok" };
+  promise_repaired => { "ok" };
+  cancel_notkept => { "ok" };
```
Comment (vpodzime) [inline tests/acceptance/15_control/02_agent/no-policy.cf]: How could this fail then?

File: tests/acceptance/15_control/02_agent/secureinput.cf [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+body perms bad_perms
+{
+    owners => { "nobody" };
+}
+
+bundle agent check
+{
+  reports:
+    "$(this.promise_filename) Pass";
```
Comment (vpodzime) [inline tests/acceptance/15_control/02_agent/secureinput.cf]: Again, how could this fail?

File: tests/unit/Makefile.am [hunk 14]
```diff
@@ -285,37 +283,34 @@ avahi_config_test_SOURCES = avahi_config_test.c \
...
 	../../cf-serverd/server_classic.c \
 	../../cf-serverd/strlist.c
-avahi_config_test_LDADD = ../../libpromises/libpromises.la libtest.la
 
 endif
 endif
 
 #file_writer_test_CPPFLAGS = -I$(top_srcdir)/libutils
-file_writer_test_SOURCES = file_writer_test.c gcov-stub.c
+file_writer_test_SOURCES = file_writer_test.c
```
Comment (vpodzime) [inline tests/unit/Makefile.am]: This doesn't seem like a related change to me.

File: libpromises/audit.c [hunk 15]
```diff
@@ -104,5 +105,6 @@ void FatalError(const EvalContext *ctx, char *s, ...)
     }
 
     EndAudit(ctx, 0);
-    exit(EXIT_FAILURE); // calling abort would bypass exit handlers and trigger subtle bugs
+    DoCleanupAndExit(EXIT_FAILURE); // calling abort would bypass exit handlers and trigger subtle bugs
+    __builtin_unreachable();
```
[the remark this answers is not in the document]
Comment (craigcomstock) [reply libpromises/audit.c]: I added that due to FatalError() being marked as NORETURN but... now I realized I should probably mark DoCleanupAndExit() as such and that will fix it in a more generic and good way. Will cleanup.

File: libpromises/mutex.c [hunk 16]
```diff
@@ -51,11 +51,9 @@ int __ThreadLock(pthread_mutex_t *mutex,
...
     if (result != 0)
     {
-        /* Log() is blocking on mutexes itself inside malloc(), so maybe not
-         * the best idea here. */
-        Log(LOG_LEVEL_ERR,
-            "Locking failure at %s:%d function %s! (pthread_mutex_lock: %s)",
-            filename, lineno, funcname, GetErrorStrFromCode(result));
+        // printf instead of Log() to avoid problems with mutexes in Log()
+        printf("error: Locking failure at %s:%d function %s! (pthread_mutex_lock: %s)",
+               filename, lineno, funcname, GetErrorStrFromCode(result));
```
[the remark this answers is not in the document]
Comment (craigcomstock) [reply libpromises/mutex.c]: Sure. I see @karlhto has a bunch of change to this in his PR: https://github.com/cfengine/core/pull/3187/files#diff-c3ee2ace174e7b59b7b33266a7a4df70R30 So I'll keep my changes simple.

File: tests/acceptance/15_control/02_agent/secureinput.cf (hunk 13 above)
Comment (craigcomstock) [reply tests/acceptance/15_control/02_agent/secureinput.cf]: It seems this is a broken test. I will work on fixing it.

File: tests/unit/Makefile.am (hunk 14 above)
Comment (craigcomstock) [reply tests/unit/Makefile.am]: Sure. I can move this to another PR. I was unable to run unit tests locally with gcov-stub.c present. Maybe it runs OK in some cloud situations since the point of gcov-stub.c is to handle the case where the codebase has gcov enabled but the test code does not.

**Comment by craigcomstock** [comment]
will see how travis/jenkins go and then squash commits for code review fixes.

**Comment by craigcomstock** [comment]
[Build Status](https://ci.cfengine.com/job/pr-pipeline/1048/)

**Comment by craigcomstock** [comment]
@cf-bottom ok, another jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1060/)

https://ci.cfengine.com/job/pr-pipeline/1060/

**Comment by craigcomstock** [comment]
@cf-bottom retry jenkins, I cleaned up my commits

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1061/)

https://ci.cfengine.com/job/pr-pipeline/1061/

**Comment by craigcomstock** [comment]
I purposely added test coverage in a separate PR which causes the codecov/patch number to be low (32.81% of diff hit (target 64.22%)). Maybe this wasn't the right approach but since I had to instrument the code a bit I thought it was a good idea.

**Comment by ncharles** [comment]
In my experience, the Windows agent is often slowed down by the lmdb read/write - files are pretty big (as sparse files are not supported) and often modified.

**Comment by craigcomstock** [comment]
Thanks @ncharles. Appreciate the feedback about slow performance of lmdb on Windows. We haven't dug into that just yet. Here we are only addressing stability especially during process exit.

**Comment by craigcomstock** [comment]
made this a WIP in hopes that a combined refactor+tests PR works better: https://github.com/cfengine/core/pull/3252

**Comment by craigcomstock** [comment]
abandoning in favor of newer PR
